### ISSUE TITLE
rounding error in theta generation

### ISSUE THREAD

**Issue by data-exchange** [tomopy/tomopy#238]
this seems to be caused by the way linspace is used in tomopy.sim.project.angles

`theta = np.linspace(0. * np.pi / 180., 180. * np.pi / 180., 1501)`

is different from:

`theta = np.linspace(0, 180, 1501) * np.pi / 180.`

the last gives a correct theta. If you agree I will correct by pull request

--decarlof as data-exchange.

**Comment by dmpelt** [comment]
I would expect these to give exactly the same results (up to machine precision), since it is all linear. What kind of issues do you get with one of them?

If I try it with a small number of angles, I indeed get:

```python
In [21]: np.linspace(0*np.pi/180.,180*np.pi/180, 5)
Out[21]: array([ 0.        ,  0.78539816,  1.57079633,  2.35619449,  3.14159265])

In [22]: np.linspace(0,180, 5)*np.pi/180
Out[22]: array([ 0.        ,  0.78539816,  1.57079633,  2.35619449,  3.14159265])
```

Even with larger numbers of angles, I get:

```python
In [25]: np.allclose(np.linspace(0,180, 1501)*np.pi/180, np.linspace(0*np.pi/180.,180*np.pi/180, 1501))
Out[25]: True
```

PS. The first version can be trivially simplified to: `np.linspace(0., np.pi, 1501)`.

**Comment by decarlof** [comment]
the error on my system is very small (-4.4408920985e-16) when using the typical angular sampling (1500 projections). This shows up when doing reconstructions using the theta generated in the two methods as a difference in the reconstructed images. I discovered this because when validating data-exchange I compare 1:1 full size reconstruction images. 

Try on your system this:

```python
import numpy as np

if __name__ == '__main__':
    # Set data collection angles as equally spaced between 0-180 degrees.
    theta_r = np.linspace(0. * np.pi / 180., 180. * np.pi / 180., 1501)
    theta_d = np.linspace(0, 180, 1501) * np.pi / 180.

    res = theta_r - theta_d
    for index in range(len(res)):
        print (res[index])
```

on my machine I get:

```python
0.0
-4.4408920985e-16
0.0
0.0
-4.4408920985e-16
4.4408920985e-16
0.0
-4.4408920985e-16
0.0
0.0
...
```

**Comment by decarlof** [comment]
or

```python
In [1]: import numpy as np
In [2]: theta = np.linspace(0. * np.pi / 180., 180. * np.pi / 180., 1501)
In [3]: theta = theta * 180. / np.pi
In [4]: print ('{0:.16f}'.format(theta[3]))
0.3599999999999999
```
versus

```python
In [5]: theta = np.linspace(0., 180., 1501.) * np.pi / 180.
In [6]: theta = theta * 180. / np.pi
In [7]: print ('{0:.16f}'.format(theta[3]))
0.3600000000000000
```

and the difference produces different reconstructed images.

**Comment by dmpelt** [comment]
These differences are indeed what I would expect as a result of the machine precision for floating point numbers. What I wouldn't expect, however, is that the reconstructed images would look (significantly) different. What kind of differences do you see in the reconstructed images? If the differences are also of machine precision size, I don't think it is possible to completely eliminate these (e.g. numpy as an [allclose](https://docs.scipy.org/doc/numpy/reference/generated/numpy.allclose.html) function specifically to compare arrays within machine precision). If the differences are larger, we probably should look into the reconstruction code to find out why it is so sensitive to the specific projection angles, and fix it there.

**Comment by decarlof** [comment]
The reconstruction difference is within the machine precision size so the issue can be ignored for real sample datasets. I noticed the difference because when I validate new data-exchange pull requests, I run a set of reconstruction against full size data sets taken at different facilities and compare the reconstructed images 1:1 with the reconstruction obtained running the master code. Some of the test data sets include theta as set by the controller (not the encoder) so it lists exact angular positions (0, 0.12, 0.24 ...) in some theta is missing and therefore generated by tomopy.  I then noticed that if theta is generated by tomopy on the same datasets the 1:1 comparison between reconstructed images fails because of the way we generate the theta (np.linspace(0. , np.pi, 1501) instead of np.linspace(0., 180., 1501)). Again the error is not a big deal for real data set but it can be annoying when using phantom and let tomopy set theta. In this case the comaprison with the ground truth will show some small error.
